### PR TITLE
Add doc generation

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @gkoz

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -884,6 +942,15 @@ impl Get for Attributes {
...
 
+fn read_text(parser: &mut Reader) -> Result<Option<String>, Error> {
+    let ret = match try!(parser.next()) {
+        Characters(text) => Ok(Some(text)),
+        _ => Ok(None),
+    };
+    try!(ignore_element(parser));
+    ret
+}
+
```
Comment (gkoz) [inline src/parser.rs]: Like I said previously I would like this function to resemble other `read_*` functions i.e. have a loop which handles `EndElement` explicitly, concatenates multiple `Characters` events if necessary and returns an error on `StartElement`.

File: src/codegen/doc/mod.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+pub fn generate(env: &Env) {
+    let path =  env.config.target_path.join("doc.cmts");
+    println!("Generating documentation {:?}", path);
+    save_to_file(&path, env.config.make_backup,
+        |w| generate_doc(w, &env.library));
+}
+
+fn generate_doc(w: &mut Write, lib: &Library) -> Result<()> {
+    for i in 0..lib.namespaces.len() {
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: Right now we only work on the "main" namespace in one generator run so just `let namespace = lib.namespaces.main();` will do.

File: src/codegen/doc/mod.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        let mut output = String::new();
+        let namespace = &lib.namespaces[i];
+
+        if let Some(ref namespace_doc) = namespace.doc {
+            for line in namespace_doc.split("\n") {
+                output.push_str(&format!("{}{}\n", FILE_COMMENT, line));
+            }
+        }
+        try!(create_sub_docs(&mut output, &namespace.constants, None));
+        for i in 0..namespace.types.len() {
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: Don't use `for i in 0..len` in Rust unless there's a good reason. This one can be handled with

```rust
for ty in ns.types.iter().filter_map(|t| t.as_ref())
```

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: Oh the `.as_ref`, I totally forgot it... And yes, it's because of the moved errors that came back to this syntax.

File: src/codegen/doc/mod.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: Ok, I'll change the code is consequence. I guess that each struct is in its own file and there isn't a namespace for one file (like I thought before).

File: src/codegen/doc/mod.rs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        w.push_str(&format!("{}{}\n{}\n",
+                            MOD_COMMENT,
+                            ty,
+                            format!("{}/// {}\n",
+                                    tabs,
+                                    class_doc.split("\n").collect::<Vec<&str>>().join(&format!("\n{}/// ", tabs)))));
+    }
+    for child_id in class.children.iter() {
+        if let &LType::Class(ref child) = lib.type_(child_id.clone()) {
+            try!(create_class_doc(w, &child, Some(Box::new(ty.clone())), lib));
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: This feels like it will generate a lot of duplicate entries. Won't `handle_type` call `create_class_doc` on all children anyway?

No `create_fn_doc`?

**Comment by EPashkin** [comment]
Can you rebase PR to master? First 2 commits was already merged.

File: src/codegen/doc/mod.rs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    if let Some(ref class_doc) = class.doc {
+        w.push_str(&format!("{}{}\n{}\n",
+                            MOD_COMMENT,
+                            ty,
+                            format!("{}/// {}\n",
+                                    tabs,
+                                    class_doc.split("\n").collect::<Vec<&str>>().join(&format!("\n{}/// ", tabs)))));
+    }
+    for child_id in class.children.iter() {
+        if let &LType::Class(ref child) = lib.type_(child_id.clone()) {
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: `TypeId` is `Copy`. It's confusing to see it `clone`d when you can `for &child_id in class.children.iter()` instead.

File: src/codegen/doc/mod.rs (hunk 4 above)
Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: It's wrong anyway, this code is based on the principle that a class can be a child of another one, which isn't the case here. `create_class_doc` function will be changed to take this in account.

File: src/codegen/doc/mod.rs (hunk 2 above)
Comment (gkoz) [inline src/codegen/doc/mod.rs]: This `Namespace` thing corresponds to a single `gir` file (a single `<namespace>` element). So `gtk` is one "namespace". When we run the generator on `gtk` it's unnecessary to generate stuff for `gdk` or `glib` etc.

**Comment by GuillaumeGomez** [comment]
@EPashkin: Done. Also, this PR isn't intended to be merged as is. It's to make the code comparison easier in order to make @gkoz task to help me easier as well. :)

**Comment by EPashkin** [comment]
Thanks.
While PR still in progress, I believe that it will be merged, so better fix sooner.

**Comment by GuillaumeGomez** [comment]
@EPashkin: Already done anyway. :p

File: src/library.rs [hunk 6]
```diff
@@ -192,20 +215,25 @@ pub struct Alias {
     pub c_identifier: String,
     pub typ: TypeId,
     pub target_c_type: String,
+    pub doc: Option<String>,
```
Comment (gkoz) [inline src/library.rs]: Would you mind doing the same with `doc-deprecated`? It is relevant for doc generation and parser-wise is exactly like `doc`.

Comment (GuillaumeGomez) [inline src/library.rs]: I'm still thinking about how I could generate it (I mean, I need to check if I can use `#[deprecated]`, otherwise I'll just make something like: "!!!!!!!!!!!! DEPRECATED STUFF !!!!!!!!!!!!!!!!!!!!!!!!".

Comment (gkoz) [inline src/library.rs]: Appending this to docs with a `# Deprecated` subheading should be workable. Not in the parser though. `#[deprecated]` is better left for some other day. ;)

Comment (GuillaumeGomez) [inline src/library.rs]: For now I have set deprecated doc to "========> DEPRECATED <=========". Should be enough. :p

File: src/codegen/doc/mod.rs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    ty.parent = parent;
+
+    try!(writeln!(w, "{}src/auto/{}.rs", FILE, class.name.to_lowercase()));
+    let found = if let Some(ref class_doc) = class.doc {
+        try!(writeln!(w, "{}{}\n{}",
+                      MOD_COMMENT,
+                      ty,
+                      format!("{}/// {}",
+                              tabs,
+                              class_doc.split("\n").collect::<Vec<&str>>().join(&format!("\n{}/// ", tabs)))));
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: There doesn't seem to be a reason to `collect` rather than `write!` each line in a `for` loop. Additionally the `found` dance will become unnecessary.

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: I avoid calling `write!` as much as possible in these cases. The script is already slow engouh.

Comment (gkoz) [inline src/codegen/doc/mod.rs]: I don't get it. Writes are buffered.

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: They do? Erf. Bad habit from C...

Comment (gkoz) [inline src/codegen/doc/mod.rs]: https://github.com/gtk-rs/gir/blob/master/src/file_saver.rs#L18

This reminds me: notice the `Untabber` filter. You can avoid messing with tab size and `"    "` because a '\t' will get expanded properly.

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: Hum... I'm gonna try.

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: I just thought that a tab won't affect the fact that:
1. it will have to be translated internally into spaces, which I do directly, so no perfomance gain
2. I still have to generate as many tabs as needed (I think one tab = 4 spaces)

There isn't much to win in here.

Comment (gkoz) [inline src/codegen/doc/mod.rs]: That's just for convenience: sometimes it's nice to not have to count spaces.

See also https://github.com/gtk-rs/gir/blob/master/src/writer/mod.rs#L6, you can import those consts instead of redefining them.

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: This, however, I'll use it!

File: src/codegen/doc/mod.rs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    for child_id in class.children.iter() {
+        if let &LType::Class(ref child) = lib.type_(*child_id) {
+            try!(create_class_doc(w, &child, None, lib));
+        }
+    }
+    Ok(())
+}
+
+fn create_trait_doc(w: &mut Write, trait_: &library::Interface, parent: Option<Box<TypeStruct>>) -> Result<()> {
+    let tabs : String = iter::repeat(" ").take(compute_ident(&parent) * TAB_SIZE).collect::<Vec<&str>>().concat();
```
Comment (EPashkin) [inline src/codegen/doc/mod.rs]: Don't `writer::primitives::tabs(num)` doing same?

File: src/library.rs [hunk 9]
```diff
@@ -192,20 +220,28 @@ pub struct Alias {
...
+    pub doc_deprecated: Option<String>,
 }
 
+
```
Comment (EPashkin) [inline src/library.rs]: Can you remove this?

File: src/library.rs [hunk 10]
```diff
@@ -192,20 +220,28 @@ pub struct Alias {
...
     pub name: String,
     pub c_identifier: String,
     pub typ: TypeId,
     pub c_type: String,
     pub value: String,
+    pub doc: Option<String>,
+    pub doc_deprecated: Option<String>,
 }
 
+
```
Comment (EPashkin) [inline src/library.rs]: And this.

File: src/config/work_mode.rs [hunk 11]
```diff
@@ -2,8 +2,9 @@ use std::str::FromStr;
 
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
 pub enum WorkMode {
-    Normal,     //generate widgets etc.
-    Sys,        //generate -sys with ffi
+    Normal,     // generate widgets etc.
+    Sys,        // generate -sys with ffi
+    Doc,        // generate documentation file
```
Comment (EPashkin) [inline src/config/work_mode.rs]: Can you add `doc` to https://github.com/gtk-rs/gir/blob/master/src/config/config.rs#L22 ?

File: src/codegen/doc/mod.rs (hunk 8 above)
Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: @gkoz: Said the same things a few comments above. And yes, I'll use it.

Comment (gkoz) [inline src/codegen/doc/mod.rs]: I'd say this is an improvement on what I pointed at because I was just talking about constants and this functions gives you the whole string at once.

**Comment by GuillaumeGomez** [comment]
I think I have adressed all comments. Anything else?

File: Cargo.lock [hunk 12]
```diff
@@ -3,14 +3,15 @@ name = "gir"
...
- "xml-rs 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)",
+ "docopt 0.6.78 (registry+https://github.com/rust-lang/crates.io-index)",
+ "env_logger 0.3.2 (registry+https://github.com/rust-lang/crates.io-index)",
+ "git2 0.3.4 (registry+https://github.com/rust-lang/crates.io-index)",
+ "lazy_static 0.1.15 (registry+https://github.com/rust-lang/crates.io-index)",
+ "log 0.3.5 (registry+https://github.com/rust-lang/crates.io-index)",
+ "regex 0.1.48 (registry+https://github.com/rust-lang/crates.io-index)",
+ "stripper_interface 0.0.1 (registry+https://github.com/rust-lang/crates.io-index)",
+ "toml 0.1.25 (registry+https://github.com/rust-lang/crates.io-index)",
+ "xml-rs 0.2.2 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (gkoz) [inline Cargo.lock]: Can the `Cargo.toml`/`Cargo.lock` update be put in a separate commit?

File: src/codegen/doc/mod.rs [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+fn create_class_doc(w: &mut Write, class: &library::Class, parent: Option<Box<TypeStruct>>, lib: &Library) -> Result<()> {
+    let tabs : String = primitives::tabs(compute_ident(&parent));
+    let mut ty = class.convert();
+    ty.parent = parent;
+
+    try!(writeln!(w, "{}src/auto/{}.rs", FILE, class.name.to_lowercase()));
+    let found = if let Some(ref class_doc) = class.doc {
+        try!(writeln!(w, "{}{}", MOD_COMMENT, ty));
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: The `found` business is unnecessary. Essentially you want to execute this line if `class_doc.is_some() && doc_deprecated.is_some()`. This can be expressed much more concisely.

File: src/codegen/doc/mod.rs [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+        if found == false {
+            try!(writeln!(w, "{}{}",
+                          MOD_COMMENT,
+                          ty));
+        }
+        try!(writeln!(w, "{}{}\n{}/// {}",
+                      tabs,
+                      DEPRECATED,
+                      tabs,
+                      class_doc.split("\n").collect::<Vec<&str>>().join(&format!("\n{}/// ", tabs))));
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: Redundant `collect/join` has been discussed already I believe.

File: src/codegen/doc/mod.rs [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+    }
+    for i in 0..class.functions.len() {
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: No C for loops please.

File: src/library.rs [hunk 16]
```diff
@@ -6,6 +6,34 @@ use nameutil::split_namespace_name;
 use traits::*;
 use version::Version;
 
+use stripper_interface as stripper;
+use stripper_interface::Type as SType;
+use stripper_interface::TypeStruct;
+
+pub trait ToStripperType {
```
Comment (gkoz) [inline src/library.rs]: This trait and its implementation for lack of better place should go to the `codegen::doc` module.

File: src/parser.rs [hunk 17]
```diff
@@ -884,6 +946,22 @@ impl Get for Attributes {
     }
 }
 
+fn read_text(parser: &mut Reader) -> Result<Option<String>, Error> {
+    match try!(parser.next()) {
```
Comment (gkoz) [inline src/parser.rs]: I realize you had been too tired when fixing this place. ;) Note how every `read_*` function contains a loop not a single read.

File: src/parser.rs [hunk 18]
```diff
@@ -884,6 +946,22 @@ impl Get for Attributes {
...
+        Characters(mut text) => {
+            Ok(if let Some(ret) = try!(read_text(parser)) {
+                text.push_str(&ret);
+                Some(text)
+            } else {
+                Some(text)
+            })
+        }
+        EndElement { .. } => Ok(None),
+        StartElement { .. } => Err(mk_error!("Unexpected StartElement", parser)),
```
Comment (gkoz) [inline src/parser.rs]: This would preferably be consistent with other "unexpected element" errors (e.g. featuring the element's name).

File: src/parser.rs [hunk 19]
```diff
@@ -884,6 +946,22 @@ impl Get for Attributes {
...
+        _ => read_text(parser),
```
Comment (gkoz) [inline src/parser.rs]: This seems inconsistent with other `read_*` functions too.

File: src/codegen/doc/mod.rs [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+use analysis::namespaces::MAIN;
+use env::Env;
+use file_saver::save_to_file;
+use library::{self, Library, ToStripperType};
+use library::Type as LType;
+use writer::primitives;
+
+use stripper_interface::{FILE, /*FILE_COMMENT, */MOD_COMMENT, TypeStruct};
+
+const DEPRECATED : &'static str = "/// ====>> DEPRECATED <<====";
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: I feel this will look more rusty as just `"/// # Deprecated"`. ;)

Comment (gkoz) [inline src/codegen/doc/mod.rs]: A '#' heading might need to be surrounded by newlines though.

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: Yes, but so much easier to see it like this. :D

**Comment by EPashkin** [comment]
Not sure if it error but now in generated docs some comments uncommented:
```
=| trait Buildable
-/// GtkBuildable allows objects to extend and customize their deserialization
+GtkBuildable allows objects to extend and customize their deserialization
```
same with `=| struct Widget` and other tabless comments

**Comment by GuillaumeGomez** [comment]
@EPashkin: No, it's abnormal.

File: src/codegen/doc/mod.rs (hunk 13 above)
Comment (gkoz) [inline src/codegen/doc/mod.rs]: And by `&&` I meant `||`... :blush:

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by EPashkin** [comment]
Now seems no "uncommented" comments.
Still not sure why output named `doc.cmts` while `rustdoc-stripper` accept only `comments.cmts`, but IMHO it not problem even if `rustdoc-stripper` can't regenerate comments now as it can be fixed later.

**Comment by GuillaumeGomez** [comment]
@EPashkin: The name of the file doesn't really matter (you can set the input file inside the rustdoc-stripper options if you want). I thought it might be better for understanding.

File: src/codegen/doc/mod.rs [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+impl_to_stripper_type!(Bitfield, Type);
+impl_to_stripper_type!(Record, Type);
+impl_to_stripper_type!(Field, Variant);
+impl_to_stripper_type!(Union, Struct);
+impl_to_stripper_type!(Function, Fn);
+impl_to_stripper_type!(Interface, Trait);
+impl_to_stripper_type!(Class, Struct);
+impl_to_stripper_type!(Namespace, Mod);
+
+const DEPRECATED : &'static str = "/// ====>> DEPRECATED <<====";
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: I don't know why the previous comment is gone, it will be more robust to add a `write_deprecated` function, which would emit
```
///
/// # Deprecated
///
```
This can't be done with just a const because of indents.

File: src/codegen/doc/mod.rs [hunk 22]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    for function in trait_.functions.iter() {
+        try!(create_fn_doc(w, &function, Some(Box::new(ty.clone()))));
+    }
+    Ok(())
+}
+
+fn create_enum_doc(w: &mut Write, enum_: &Enumeration, parent: Option<Box<TypeStruct>>) -> Result<()> {
+    let ident = compute_ident(&parent);
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: Looks like `s/ident/indent/g` throughout the module is in order.

File: src/codegen/doc/mod.rs [hunk 23]
```diff
@@ -0,0 +1,255 @@
...
+        }
+    }
+    Ok(())
+}
+
+fn create_fn_doc(w: &mut Write, fn_: &Function, parent: Option<Box<TypeStruct>>) -> Result<()> {
+    let tabs : String = primitives::tabs(compute_ident(&parent) + 1);
+
+    let mut docs = match fn_.doc {
+        Some(ref d) => format!("{}/// {}\n", tabs, d.split("\n").collect::<Vec<&str>>().join(&format!("\n{}/// ", tabs))),
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: I'd prefer this too was done straightforwardly without collecting into `docs` (not to mention that this function bypasses `write_lines`).
At worst you'll have empty `MOD_COMMENT` line for a couple of functions, which should not have a noticeable effect on stripper.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/codegen/doc/mod.rs [hunk 24]
```diff
@@ -0,0 +1,271 @@
...
+    for ty in namespace.types.iter().filter_map(|t| t.as_ref()) {
+        try!(handle_type(&mut w, &ty, None, lib));
+    }
+    Ok(())
+}
+
+fn handle_type(w: &mut Write, ty: &LType, parent: Option<Box<TypeStruct>>, lib: &Library) -> Result<()> {
+    match *ty {
+        LType::Alias(ref a) => {
+            try!(writeln!(w, "{}src/auto/{}.rs", FILE, a.name.to_lowercase()));
```
Comment (EPashkin) [inline src/codegen/doc/mod.rs]: `to_lowercase()` not right function, better use `nameutil::module_name`

Comment (EPashkin) [inline src/codegen/doc/mod.rs]: It convert `ActionBar` to `action_bar`

Comment (gkoz) [inline src/codegen/doc/mod.rs]: This is going to be fun because `GLArea` and GLContext` need module name overrides from config. We could leave this off for another PR though.

**Comment by EPashkin** [comment]
May be problem that files in `doc.cmts` placed in random order, but only if this file saved.
Still don't understand reason for randomness as `generate_doc` iterates `types: Vec<Option<Type>>`

File: src/codegen/doc/mod.rs [hunk 25]
```diff
@@ -0,0 +1,271 @@
...
+            try!(write_lines(w, &m_doc, &tabs));
+        }
+    }
+    Ok(())
+}
+
+fn create_fn_doc(w: &mut Write, fn_: &Function, parent: Option<Box<TypeStruct>>) -> Result<()> {
+    let tabs : String = primitives::tabs(compute_indent(&parent) + 1);
+
+    if fn_.doc().is_none() && fn_.doc_deprecated().is_none() && fn_.ret.doc().is_none() && fn_.ret.doc_deprecated().is_none() {
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: This doesn't fit on the github page it probably is longer than 100 chars.

File: src/codegen/doc/mod.rs (hunk 24 above)
Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: Yes, there are some cases where the file name doesn't correspond to what is inside.

File: src/codegen/doc/mod.rs [hunk 26]
```diff
@@ -0,0 +1,271 @@
...
+        let mut leave = true;
+
+        for parameter in fn_.parameters.iter() {
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: `parameters.iter().any()`?

File: src/codegen/doc/mod.rs [hunk 27]
```diff
@@ -0,0 +1,271 @@
...
+        try!(write_deprecated(w, &tabs));
+        try!(write_lines(w, &class_doc, &tabs));
+    }
+
+    for function in class.functions.iter() {
+        try!(create_fn_doc(w, &function, Some(Box::new(ty.clone()))));
+    }
+    for child_id in class.children.iter() {
+        if let &LType::Class(ref child) = lib.type_(*child_id) {
+            try!(create_class_doc(w, &child, None, lib));
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: Why is `create_class_doc` recursive? I've now confirmed my suspicion that it generates redundant copies of documentation.

**Comment by gkoz** [comment]
It looks like you can't get away with not relying on analysis. But the good news is that a good bit of classes handling can be copied from https://github.com/gtk-rs/gir/blob/master/src/codegen/objects.rs#L10-L30

Here instead of taking the classes and interfaces from the library you iterate over "objects" from the config and take ones, that are not "ignore"d. This computes the correct module name and `Info::has_children` tells you if there is a trait or just a struct. Then functions, which have `instance_parameter = true` parameters go into traits.

**Comment by EPashkin** [comment]
IMHO better don't modify iteration and check in `create_class_doc` with `env::type_status` and next do analysis

**Comment by EPashkin** [comment]
Was wrong: `env::type_status` is not enough as also needed configures `module_name`, so
instead use something like `config.objects.get(name).unwrap_or(Default::default())`

**Comment by GuillaumeGomez** [comment]
@EPashkin: How could this work? `name` can have conflicts, how does it differentiate between two indentical names?

**Comment by gkoz** [comment]
That `name` would be a "full name" e.g. `Gtk.Button` but I don't think you'll have it handy, so you'll need to format it but then this approach becomes somewhat cumbersome.

**Comment by GuillaumeGomez** [comment]
If I have to do any transformation on it, then I don't think it's a good idea.

**Comment by EPashkin** [comment]
`name` is something like 'Gtk.GLArea' can be obtained with `type_id.full_name(&env.library)`.
You right about conflicting names, contained in `library::Class`, so you need pass to `handle_type` `namespace.name' or enumerate iterator and combine TypeId with MAIN.

**Comment by EPashkin** [comment]
IMHO skipping ignored can be added later. I just commented @gkoz idea about iterating config.

**Comment by EPashkin** [comment]
On second though iterating config is better that iterating library.

**Comment by GuillaumeGomez** [comment]
Ok, should be it.

File: src/codegen/doc/mod.rs [hunk 28]
```diff
@@ -0,0 +1,296 @@
...
+           x.doc().is_some() || x.doc_deprecated().is_some()
+       }) {
+        try!(writeln!(w, "{}///", tabs));
+        try!(writeln!(w, "{}/// # Parameters", tabs));
+        try!(writeln!(w, "{}///", tabs));
+    }
+    let mut wrote = false;
+    for parameter in fn_.parameters.iter() {
+        if let Some(ref parameter_doc) = parameter.doc() {
+            try!(writeln!(w, "{}/// ## {}:", tabs, parameter.name));
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: This is where `instance_parameter` is relevant too. Most likely it should be skipped because in Rust it's passed implicitly.

Comment (gkoz) [inline src/codegen/doc/mod.rs]: Anyway there are endless issues with parameters so refining that can be left for later.

Comment (EPashkin) [inline src/codegen/doc/mod.rs]: Yes it really endless issues, at minimum we have out parameters to return value conversion.

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: I skip `instance_parameter` now as well.

Comment (EPashkin) [inline src/codegen/doc/mod.rs]: Strange but on `Gir_Gtk.toml` only `# Parameters` found with `## error:` with out real parameter docs.

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: Yes, that's what I have too.

File: src/codegen/doc/mod.rs [hunk 29]
```diff
@@ -0,0 +1,296 @@
...
+        |w| generate_doc(w, &env));
+}
+
+fn generate_doc(mut w: &mut Write, env: &Env) -> Result<()> {
+    //let namespace = lib.namespace(MAIN);
+
+    //try!(create_sub_docs(&mut output, &namespace.constants, None));
+    /*for ty in namespace.types.iter().filter_map(|t| t.as_ref()) {
+        try!(handle_type(&mut w, &ty, None));
+    }*/
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: Iterating over `config.objects` is only suitable for objects but the rest (enums, records, etc.) still has to be done by iterating the library, so this loop is still needed, just remove `Class` and `Interface` from `handle_type`.

File: src/codegen/doc/mod.rs [hunk 30]
```diff
@@ -0,0 +1,296 @@
...
+        LType::Function(ref f) => create_fn_doc(w, &f, parent),
+        LType::Interface(ref i) => create_class_doc(w, i, parent, is_trait),
+        LType::Class(ref c) => create_class_doc(w, c, parent, is_trait),
+        _ => Ok(()),
+    }
+}
+
+fn create_class_doc<T: FunctionTraitType + ToStripperType>(w: &mut Write, class: &T,
+                                                           parent: Option<Box<TypeStruct>>,
+                                                           is_trait: bool)
```
Comment (gkoz) [inline src/codegen/doc/mod.rs]: `is_trait` is misleading because its meaning is `has_trait`

File: src/codegen/doc/mod.rs (hunk 29 above)
Comment (EPashkin) [inline src/codegen/doc/mod.rs]: All that have own file count as object in config.

Comment (GuillaumeGomez) [inline src/codegen/doc/mod.rs]: So what do I do?

Comment (gkoz) [inline src/codegen/doc/mod.rs]: I exepct the enums and bitfields to be missing from configs for a while. But that could be discussed I guess.
In the meantime we'd still like to document them even if it's right in the sys crates where they're now.

Comment (EPashkin) [inline src/codegen/doc/mod.rs]: I suggest merge the PR and then finish the remaining issues.

File: src/parser.rs [hunk 31]
```diff
@@ -884,6 +946,27 @@ impl Get for Attributes {
...
+        match try!(parser.next()) {
+            Characters(text) => {
+                ret_text = match ret_text {
+                    Some(t) => Some(format!("{}{}", t, text)),
+                    None => Some(text),
+                }
+            }
+            EndElement { .. } => break,
+            StartElement { name, .. } => return Err(mk_error!(&format!("Unexpected element: {}",
+                                                                       name),
```
Comment (gkoz) [inline src/parser.rs]: This is materially different from how other unexpected element errors are generated. `name` is not the same as `name.local_name`.

Comment (GuillaumeGomez) [inline src/parser.rs]: I don't see anything like this :p

File: src/parser.rs [hunk 32]
```diff
@@ -884,6 +946,27 @@ impl Get for Attributes {
...
+                ret_text = match ret_text {
+                    Some(t) => Some(format!("{}{}", t, text)),
+                    None => Some(text),
+                }
+            }
+            EndElement { .. } => break,
+            StartElement { name, .. } => return Err(mk_error!(&format!("Unexpected element: {}",
+                                                                       name.local_name),
+                                                              parser)),
+            _ => continue,
```
Comment (gkoz) [inline src/parser.rs]: This is not how other `read_*` functions handle the catch-all case.

Comment (GuillaumeGomez) [inline src/parser.rs]: Sorry, fixed.

**Comment by gkoz** [comment]
Yeah looks good. I think it's time to sqash most in-progress commits (not the `Cargo.*` one).

**Comment by GuillaumeGomez** [comment]
Squashed enough?

**Comment by gkoz** [comment]
Yeah.

My testing doesn't show any parameter documentation but I don't see an obvious problem in the code.

**Comment by GuillaumeGomez** [comment]
That's rassuring and scary at the same time. I wasn't able to determine what I have missed.

**Comment by EPashkin** [comment]
@GuillaumeGomez parameters docs was just not parsed and error get functions docs.
Fix https://github.com/EPashkin/gir/commit/7ef8557c3bad696b8c48bb8af4a270245128860c
IMHO still need some empty line before `# returns`.
And output parameters also need some change but it better do later.

**Comment by EPashkin** [comment]
IMHO also better delete `Parameter::doc_deprecated` and `impl_to_stripper_type!(Parameter, Variant);` or at minimum delete `if let Some(ref parameter_doc) = parameter.doc_deprecated() {` that I missed.

**Comment by GuillaumeGomez** [comment]
@EPashkin: Why that?

**Comment by EPashkin** [comment]
@GuillaumeGomez sorry don't understand. No 'doc_deprecated' in 'parameters' and 'return-value`

**Comment by GuillaumeGomez** [comment]
@EPashkin: Oh ok! I update my code then.

**Comment by GuillaumeGomez** [comment]
@EPashkin: I rebased my code on your fix but I still don't have parameters. I also removed doc deprecated generation for parameters and return value.

**Comment by EPashkin** [comment]
3 changes you don't included https://gist.github.com/EPashkin/be4a162bb75dab540ef8

**Comment by EPashkin** [comment]
May be rebase way was wrong, as my comment was on your previous.

**Comment by GuillaumeGomez** [comment]
My PR is 3 commits above gir/master. So are you sure you did include it correctly?

**Comment by EPashkin** [comment]
Change I gisted is between your last 3c3a81f3c3803e45d5c5c8afa35fcfd36e02afd9 and "good" code.
For some reason your rebase don't get needed changes, so try do it manually.

**Comment by GuillaumeGomez** [comment]
@EPashkin: That's an interesting bug. I added the changes. Weird that the rebase failed...

**Comment by EPashkin** [comment]
Now present unnamed parameter from varargs addition in line parser.rs:882.
IMHO better set it to `None` as no varargs in Rust.
```
=| struct InfoBar§fn new_with_buttons
    /// Creates a new #GtkInfoBar with buttons. Button text/response ID
    /// pairs should be listed, with a %NULL pointer ending the list.
    /// Button text can be either a stock ID such as %GTK_STOCK_OK, or
    /// some arbitrary text. A response ID can be any positive number,
    /// or one of the values in the #GtkResponseType enumeration. If the
    /// user clicks one of these dialog buttons, GtkInfoBar will emit
    /// the “response” signal with the corresponding response ID.
    ///
    /// # Parameters
    ///
    /// ## first_button_text:
    /// stock ID or text to go in first button, or %NULL
    /// ## :
    /// response ID for first button, then additional buttons, ending
    ///    with %NULL
```

**Comment by GuillaumeGomez** [comment]
@EPashkin: If name is empty, I don't display this parameter. Does it seem fine to you?

**Comment by EPashkin** [comment]
Now docs equal my previous.

**Comment by GuillaumeGomez** [comment]
@EPashkin: What do you mean? I still have parameters.

**Comment by EPashkin** [comment]
I mean is good, change successfully merged and result equal my "proposed"

**Comment by GuillaumeGomez** [comment]
Perfect then. :)

File: src/codegen/doc/mod.rs [hunk 33]
```diff
@@ -0,0 +1,283 @@
...
+    if let Some(ref docs) = fn_.doc() {
+        try!(write_lines(w, &docs, &tabs));
+    };
+    if let Some(ref docs) = fn_.doc_deprecated() {
+        try!(write_header(w, &tabs, "Deprecated"));
+        try!(write_lines(w, &docs, &tabs));
+    };
+
+    if fn_.parameters.iter().any(|x| {
+           x.instance_parameter == false && x.name.len() > 0 && x.doc().is_some()
```
Comment (EPashkin) [inline src/codegen/doc/mod.rs]: Not better use `String::is_empty` instead `len`

File: src/codegen/doc/mod.rs [hunk 34]
```diff
@@ -0,0 +1,283 @@
...
+       }) {
+        try!(write_header(w, &tabs, "Parameters"));
+    }
+    for parameter in fn_.parameters.iter() {
+        if parameter.instance_parameter || parameter.name.len() < 1 {
```
Comment (EPashkin) [inline src/codegen/doc/mod.rs]: Same

**Comment by gkoz** [comment]
Thanks for contributing to making nice documentation possible!

**Comment by GuillaumeGomez** [comment]
@gkoz: Thanks for your great help! @EPashkin: And thanks a lot to you as well! :)
